drivers/flash/spi_nand.c

```diff
@@ -0,0 +1,1038 @@
+			 const void *src,
+			 size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+	uint32_t offset = 0;
+	uint32_t chunk = 0;
+	uint32_t written_bytes = 0;
+
+	/* address must be sub-page-aligned */
+	if (!(addr & (SPI_NAND_SUB_PAGE_SIZE - 1)) && (addr != 0)) {
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sub-page */
+	if ((size % SPI_NAND_SUB_PAGE_SIZE) != 0) {
+		return -EINVAL;
+	}
+
+	acquire_device(dev);
+
+	while (size > 0) {
+
+		/* Don't write more than a page. */
+		offset = addr % data->page_size;
+		chunk = (offset + size < data->page_size) ? size : (data->page_size - offset);
+		written_bytes = chunk;
+
+		spi_nand_cmd_write(dev, SPI_NAND_CMD_WREN);
+
+		if (data->ecc_bits == 0) {
+			ret = spi_nand_access(dev, SPI_NAND_CMD_PP_LOAD,
+				NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_16BIT_ADDR,
+				(addr & SPI_NAND_PAGE_MASK), (void *)src, written_bytes);
+			if (ret != 0) {
+				LOG_ERR("program load failed: %d", ret);
+				goto out;
+			}
+		} else {
+			uint8_t *p = (uint8_t *)data->page_buf;
+			uint8_t ecc_steps = data->ecc_steps;
+
+			if (size < data->page_size) {
+				LOG_ERR("Write failed");
+				return -EINVAL;
+			}
+
+			/* prepare data */
+			memset(data->page_buf, 0xff, data->page_size + data->oob_size);
+			memcpy(data->page_buf + offset, (uint8_t *)src, written_bytes);
+
+			/* calculate the software ECC */
+			for (uint8_t i = 0; ecc_steps; ecc_steps--, i += data->ecc_bytes, p += data->ecc_size) {
+				memset(data->nbc.bch->input_data, 0x0, (1 << data->nbc.bch->m) / 8);
+				memcpy(data->nbc.bch->input_data + data->ecc_bytes, p, data->ecc_size);
+				bch_calculate_ecc(dev, data->nbc.bch->input_data, data->ecc_calc + i);
+			}
+
+			/* prepare ECC code */
+			memcpy(data->page_buf + data->page_size + data->ecc_layout_pos,
+				data->ecc_calc, data->ecc_bytes * data->ecc_steps);
+
+			written_bytes = data->page_size + data->oob_size;
+			ret = spi_nand_access(dev, SPI_NAND_CMD_PP_LOAD,
+				NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_16BIT_ADDR,
+				(addr & SPI_NAND_PAGE_MASK), (void *)data->page_buf, written_bytes);
+			if (ret != 0) {
+				LOG_ERR("program load failed: %d", ret);
+				goto out;
+			}
+		}
+
+		ret = spi_nand_access(dev, SPI_NAND_CMD_PROGRAM_EXEC, 
+			NAND_ACCESS_WRITE | NAND_ACCESS_ADDRESSED | NAND_ACCESS_24BIT_ADDR,
+			addr >> data->page_shift, NULL, 0);
+		if (ret != 0) {
+			LOG_ERR("program excute failed: %d", ret);
+			goto out;
+		}
+
+		src = (const uint8_t *)(src) + chunk;
+		addr = (addr + SPI_NAND_PAGE_OFFSET) & (~SPI_NAND_PAGE_MASK);
+		size -= chunk;
+
+		ret = spi_nand_wait_until_ready(dev);
+	}
+
+out:	release_device(dev);
+	return ret;
+}
+
+static int spi_nand_erase(const struct device *dev, off_t addr, size_t size)
+{
+	struct spi_nand_data *data = dev->data;
+	int ret = 0;
+
+	/* address must be block-aligned */
+	if (!(addr & SPI_NAND_BLOCK_MASK) && (addr != 0)) {
```

Comment:
`if (addr & ~SPI_NAND_BLOCK_MASK)`, the if seems complicated.

Reply:
Do you mean to change !(addr & SPI_NAND_BLOCK_MASK) to (addr & ~SPI_NAND_BLOCK_MASK)?
@de-nordic

Reply:
if (!(addr & SPI_NAND_BLOCK_MASK) && (addr != 0)) contains an error. 

You should remove the first logical not ("!") to correct it.

result: 
if ((addr & SPI_NAND_BLOCK_MASK) && (addr != 0))

!(addr & SPI_NAND_BLOCK_MASK) && (addr != 0) 
returns true on address 0, and on any page size that is a multiple of the block size () (which shouldn't happen)
i.e. address 262144 (byte 0 of page 0 of block 1) returns true and returns an error (also shouldn't happen

Reply:
could just be:
if (addr & SPI_NAND_BLOCK_MASK)